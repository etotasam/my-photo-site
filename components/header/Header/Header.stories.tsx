import { Header, HeaderType } from "./Header";
import { Story, Meta } from "@storybook/react";

export default {
  title: "Header",
  component: Header,
} as Meta<typeof Header>;
// const Caution = () => {
//   return <p style={{ marginTop: "50px" }}>controlsのisModalActiveのtrue/falseでアニメーションを確認</p>;
// };

const Template: Story<typeof Header> = (args: any) => {
  return (
    <>
      <Header {...args} />
    </>
  );
};
export const DefaultHeader = Template.bind({});
DefaultHeader.args = {
  device: "PC",
  locationNames: ["てすと1", "てすと2"],
  photoLabelName: "てすと1",
  toggleModal: () => {},
  isModalActive: false,
  // headerRef: (<div>test</div>) as unknown,
} as HeaderType;
