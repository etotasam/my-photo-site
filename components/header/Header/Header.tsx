import { useEffect, useState, useRef } from "react";
import { useRouter } from "next/router";
import { AnimatePresence } from "framer-motion";
import useSWR from "swr";
import axios from "axios";
import { HeaderNavOnMobile } from "../HeaderNavOnMobile";
import { HeaderNavOnPC } from "../HeaderNavOnPC";
import { MainModal } from "../MainModal";
import { ModalState, useModalStateContext, useModalDispatchContext } from "@/context/modalStateContext";
import { useHeihgtDispatchContext } from "@/context/heightStateContext";
import { useWindowResize } from "@/hooks";

export const Header = () => {
  const router = useRouter();
  const { isModalActive }: ModalState = useModalStateContext();
  const { modalOpenDispathcer, modalCloseDispatcher } = useModalDispatchContext();
  const { setHeaderHeightDispatcher } = useHeihgtDispatchContext();

  const breakpointWidth = 768;
  const apiUrl = process.env.API_URL;
  const fetcher = async (url: string) => await axios.get(url).then((res) => res.data);
  const { data: locations, error } = useSWR(`${apiUrl}/locations`, fetcher);

  // viewportの幅からモバイルかどうかを判断
  const { width: windowWidth } = useWindowResize();
  const isMobile = windowWidth < breakpointWidth;
  useEffect(() => {
    if (isModalActive) return modalCloseDispatcher();
  }, [isMobile]);

  const handleClick = (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
    e.preventDefault();
    modalCloseDispatcher();
    if (router.pathname === "/") return;
    router.push(`/`);
    return;
  };

  // set <header> height to HeadersContext
  const ref = useRef<HTMLElement>(null);
  useEffect(() => {
    if (!ref.current) return;
    const headerHeight: number = ref.current.clientHeight;
    setHeaderHeightDispatcher(headerHeight);
  }, [ref]);

  return (
    <header
      ref={ref}
      className={`t-header-height bg-white/90 fixed flex justify-center top-0 left-0 w-full md:w-[99vw] z-50 duration-300 ${
        isModalActive ? `bg-opacity-100` : `bg-opacity-90`
      }`}
    >
      <div className={`flex relative items-center w-[90%] max-w-[1024px] mx-auto`}>
        {windowWidth === 0 ? (
          ``
        ) : isMobile ? (
          <HeaderNavOnMobile />
        ) : (
          isMobile !== undefined && <HeaderNavOnPC locations={locations} error={error} />
        )}
        <div className={`absolute right-0`}>
          <a
            href={`/`}
            className={`text-green-600 text-xl tracking-wider font-extralight`}
            onClick={(e) => handleClick(e)}
          >
            {process.env.NEXT_PUBLIC_SITE_TITLE}
          </a>
        </div>
      </div>
      <AnimatePresence>{isModalActive && <MainModal locations={locations} error={error} />}</AnimatePresence>
    </header>
  );
};
