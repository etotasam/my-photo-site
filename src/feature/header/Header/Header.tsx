import { useRef, useEffect } from "react";
import { AnimatePresence } from "framer-motion";
import clsx from "clsx";
//! component
import { NavOnSP, NavOnSPType } from "../component/NavOnSP";
import { NavOnPC, NavOnPCType } from "../component/NavOnPC";
import { NavMenuModal } from "../component/NavMenuModal";

export type HeaderType = {
  device: "PC" | "SP" | undefined;
  handleClick: (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void;
  headerRef: React.RefObject<HTMLElement>;
} & NavOnSPType &
  NavOnPCType;

export const Header = (props: HeaderType) => {
  //? Header props
  const { device, handleClick, headerRef } = props;
  //? NavOnSP props
  const { toggleModal, isModalActive } = props;
  //? NavOnPC props
  const { locationNames, toLink, imagesLocationNamesOnRouterQuery } = props;
  //? MainModal props
  const {
    locationNames: locationNamesForModal,
    imagesLocationNamesOnRouterQuery: imagesLocationNamesOnRouterQueryForModal,
  } = props;

  return (
    <header
      ref={headerRef}
      className={clsx(
        `t-header-height fixed flex justify-center top-0 left-0 w-[100vw] z-30 duration-300 bg-white/90`
        // isModalActive ? `bg-transparent` : `bg-white/90`
      )}
    >
      <div className={`flex relative items-center w-[95vw] sm:w-[90vw] max-w-[1024px] mx-auto`}>
        {device === "SP" && <NavOnSP isModalActive={isModalActive} toggleModal={toggleModal} />}
        {device === "PC" && (
          <NavOnPC
            locationNames={locationNames.sort()}
            toLink={toLink}
            imagesLocationNamesOnRouterQuery={imagesLocationNamesOnRouterQuery}
          />
        )}
        {/* //? title */}
        <div className={`absolute right-3`}>
          <a
            href={`/`}
            className={clsx(
              "text-xl tracking-wider font-extralight",
              isModalActive ? `text-white/90` : `text-green-600`
            )}
            onClick={(e) => handleClick(e)}
          >
            {process.env.NEXT_PUBLIC_SITE_TITLE}
          </a>
        </div>
      </div>
      {/* //? modal */}
      <AnimatePresence>
        {isModalActive && (
          <NavMenuModal
            locationNames={locationNamesForModal.sort()}
            imagesLocationNamesOnRouterQuery={imagesLocationNamesOnRouterQueryForModal}
          />
        )}
      </AnimatePresence>
    </header>
  );
};
