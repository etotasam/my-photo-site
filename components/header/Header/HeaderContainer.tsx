import React, { useEffect, useRef } from "react";
import { useRouter } from "next/router";
//! hooks
import { useDeviceCheck } from "@/hooks";
//! component

import { Header } from "./Header";
//! context
import { useLocationNamesStateContext } from "@/context/locationNamesContext";
import { useModalStateContext, useModalDispatchContext } from "@/context/modalStateContext";
import { useHeihgtDispatchContext } from "@/context/heightStateContext";

export const HeaderContainer = () => {
  //! NavOnPC logic
  const router = useRouter();
  const photoLabelName = router.query.photo_label;

  const { locationNames } = useLocationNamesStateContext();

  const toLink = (location: string, e: React.MouseEvent<HTMLAnchorElement>) => {
    e.preventDefault;
    if (location === photoLabelName) return;
    router.push(`/photo/${location}?image=1`);
  };

  //! NavOnSP logic
  const { isModalActive } = useModalStateContext();
  const { modalOpenDispathcer, modalCloseDispatcher } = useModalDispatchContext();
  //? modalの開閉
  const toggleModal = () => {
    if (isModalActive) return modalCloseDispatcher();
    modalOpenDispathcer();
  };

  //! header logic
  const { device } = useDeviceCheck();

  //? deviceがPCの時はmodalを閉じる
  useEffect(() => {
    if (device === "PC") return modalCloseDispatcher();
  }, [device]);

  const handleClick = (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
    e.preventDefault();
    modalCloseDispatcher();
    if (router.pathname === "/") return;
    router.push(`/`);
    return;
  };

  const { setHeaderHeightDispatcher } = useHeihgtDispatchContext();
  const headerRef = useRef<HTMLElement>(null);
  useEffect(() => {
    if (!headerRef.current) return;
    const headerHeight: number = headerRef.current.clientHeight;
    setHeaderHeightDispatcher(headerHeight);
  }, [headerRef.current]);

  return (
    <Header
      locationNames={locationNames}
      device={device}
      photoLabelName={photoLabelName}
      toLink={toLink}
      isModalActive={isModalActive}
      toggleModal={toggleModal}
      handleClick={handleClick}
      headerRef={headerRef}
    />
  );
};
