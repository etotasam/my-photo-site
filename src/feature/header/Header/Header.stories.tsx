import { Header, HeaderType } from "./Header";
import { Story, Meta } from "@storybook/react";
import { action } from "@storybook/addon-actions";

export default {
  title: "Header",
  component: Header,
} as Meta<typeof Header>;

const handleClick = action("テスト");

const locationNames = ["egypt", "turkey"];

const Template: Story<typeof Header> = (args: any) => {
  return (
    <>
      <Header {...args} />
    </>
  );
};
export const Default = Template.bind({});
Default.args = {
  device: "PC",
  locationNames,
  imagesLocationNamesOnRouterQuery: locationNames[0],
  toggleModal: action("toggleModalFunc"),
  isModalActive: false,
  toLink: action("toLink"),
  handleClick: (e) => {
    e.preventDefault();
    handleClick(e);
  },
} as HeaderType;
