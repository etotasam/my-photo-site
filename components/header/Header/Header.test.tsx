import React from "react";
import { render, screen, act } from "@testing-library/react";
import { Header } from "./Header";
import { useRouter } from "next/router";

jest.mock(`next/router`, () => ({
  useRouter() {
    return {
      query: {},
    };
  },
}));

describe(`Header`, () => {
  it(`windowのwidthのサイズで表示されるコンポーネントが変わる`, () => {
    const { asFragment, rerender } = render(<Header />);
    act(() => resizeWidth(769));
    rerender(<Header />);
    expect(screen.queryByTestId(`pc`)).toBeInTheDocument();
    expect(screen.queryByTestId(`humburger`)).toBeNull();
    expect(asFragment()).toMatchSnapshot();
    act(() => resizeWidth(767));
    rerender(<Header />);
    expect(screen.queryByTestId(`humburger`)).toBeInTheDocument();
    expect(screen.queryByTestId(`pc`)).toBeNull();
  });
});

const resizeWidth = (num: number) => {
  window.innerWidth = num;
  window.dispatchEvent(new Event("resize"));
};
