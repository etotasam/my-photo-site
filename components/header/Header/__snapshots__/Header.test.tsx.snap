// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header windowのwidthのサイズで表示されるコンポーネントが変わる 1`] = `
<DocumentFragment>
  <header
    class="t-header-height bg-white/90 fixed flex justify-center top-0 left-0 w-full md:w-[99vw] z-50 duration-300 bg-opacity-90"
  >
    <div
      class="flex relative items-center w-[90%] max-w-[1024px] mx-auto"
    >
      <ul
        class="flex"
        data-testid="pc"
      />
      <div
        class="absolute right-0"
      >
        <a
          class="text-green-600 text-xl tracking-wider font-extralight"
          href="/"
        />
      </div>
    </div>
  </header>
</DocumentFragment>
`;
